aws/security-group/security-group.tf

```diff
@@ -77,137 +77,169 @@ resource "aws_security_group" "this" {
   type                     = "ingress"
   source_security_group_id = each.value.sg_id
 
-  description = var.rules[each.value.ingress_rules[0]][3]
-  from_port   = var.rules[each.value.ingress_rules[0]][0]
-  to_port     = var.rules[each.value.ingress_rules[0]][1]
-  protocol    = var.rules[each.value.ingress_rules[0]][2]
+  description = each.value.description
+  from_port   = each.value.from_port
+  to_port     = each.value.to_port
+  protocol    = each.value.ip_protocol
 }
 
 ############################################
 # Egress Rules - CIDR Blocks
 ############################################
-resource "aws_security_group_rule" "egress_cidr" {
+resource "aws_vpc_security_group_egress_rule" "egress_cidr" {
   for_each = var.create_default_security_group ? {} : {
-    for sg_name, sg in var.security_groups : sg_name => {
-      sg_name      = sg.name
-      sg_id        = aws_security_group.this[sg.name].id
-      egress_rules = sg.egress_rules
-      cidr_blocks  = sg.egress_cidr_blocks
-    } if length(sg.egress_cidr_blocks) > 0
+    for sg_name, sg in var.security_groups :
+    sg_name => flatten([
+      for ingress_rule in sg.ingress_rules : {
+        sg_name     = sg.name
+        sg_id       = aws_security_group.this[sg.name].id
+        description = var.rules[ingress_rule][3]
+        from_port   = var.rules[ingress_rule][0]
+        to_port     = var.rules[ingress_rule][1]
+        ip_protocol = var.rules[ingress_rule][2]
+        cidr_blocks = sg.egress_cidr_blocks
+      }
+    ]) if length(sg.egress_cidr_blocks) > 0
   }
 
   security_group_id = each.value.sg_id
-  type              = "egress"
-  cidr_blocks       = each.value.cidr_blocks
+  cidr_ipv4         = each.value.cidr_blocks
 
-  description = var.rules[each.value.egress_rules[0]][3]
-  from_port   = var.rules[each.value.egress_rules[0]][0]
-  to_port     = var.rules[each.value.egress_rules[0]][1]
-  protocol    = var.rules[each.value.egress_rules[0]][2]
+  description = each.value.description
+  from_port   = each.value.from_port
+  to_port     = each.value.to_port
+  ip_protocol = each.value.ip_protocol
 }
 
 ############################################
 # Egress Rules - IPv6 CIDR Blocks
 ############################################
-resource "aws_security_group_rule" "egress_ipv6" {
+resource "aws_vpc_security_group_egress_rule" "egress_ipv6" {
   for_each = var.create_default_security_group ? {} : {
-    for sg_name, sg in var.security_groups : sg_name => {
-      sg_name          = sg.name
-      sg_id            = aws_security_group.this[sg.name].id
-      egress_rules     = sg.egress_rules
-      ipv6_cidr_blocks = sg.egress_ipv6_cidr_blocks
-    } if length(sg.egress_ipv6_cidr_blocks) > 0
+    for sg_name, sg in var.security_groups :
+    sg_name => flatten([
+      for ingress_rule in sg.ingress_rules : {
+        sg_name     = sg.name
+        sg_id       = aws_security_group.this[sg.name].id
+        description = var.rules[ingress_rule][3]
```

Comment:
Could we use an object for rules instead of a list, it would be more readable and understandable